feat: P-701 add BRC20 contracts

wli-pro opened the pull request:
This PR implemented the token holding amount VC, with BRC20 token firstly.
It also added dependent precompile contracts: string_to_number, identity_to_string

Each token contract will be deployed to as one contract for client use. The deployed contract id will be used in requestVc call from client side, to indicate that this solidity contract will be executed to generated the VC

### Context

<!-- Why are these changes needed? Please use auto-close keyword to link to an existing issue, if any -->

### Labels

Please apply following PR-related labels when appropriate:
- `C0-breaking`: if your change could break the existing client, e.g. API change, critical logic change 
- `C1-noteworthy`: if your change is non-breaking, but is still worth noticing for the client, e.g. reference code improvement

### How (Optional)

<!-- How were these changes implemented? -->

### Testing Evidences

Please attach any relevant evidences if applicable

Case 1 (true VC):
```
$ ./bin/litentry-cli -U wss://localhost/ trusted -d link-identity did:litentry:substrate:0x16901bd268d852212935c304641228b8cee8e84a7e830faa391aba00cf71231c did:litentry:bitcoin:0x02e8c39e82aaaa143c3def8d3c7084a539b227244ac9067c3f7fc86cb73a0b7aed bitcoinp2tr
$ ./bin/litentry-cli -U wss://localhost/ trusted -d request-vc did:litentry:substrate:0x16901bd268d852212935c304641228b8cee8e84a7e830faa391aba00cf71231c -a "dynamic 0000000000000000000000000000000000000001"
>>> identity: Substrate(Address32)
>>> nonce: 1
>>> assertions: [Dynamic(0x0000000000000000000000000000000000000001)]
```
----Generated VC-----
```
{
   "@context":[
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1"
   ],
   "id":"0x33f98a2d179f2d0f312f0a2204894e8afb0ab5e5a92d525d1f2da282fde5a4b9",
   "type":[
      "VerifiableCredential"
   ],
   "credentialSubject":{
      "id":"did:litentry:substrate:0x16901bd268d852212935c304641228b8cee8e84a7e830faa391aba00cf71231c",
      "description":"The amount of a particular token you are holding",
      "type":"Token Holding Amount",
      "assertionText":"Dynamic(0x0000000000000000000000000000000000000001)",
      "assertions":[
         {
            "and":[
               {
                  "src":"$token",
                  "op":"==",
                  "dst":"ordi"
               },
               {
                  "src":"$holding_amount",
                  "op":">=",
                  "dst":"100"
               },
               {
                  "src":"$holding_amount",
                  "op":"<",
                  "dst":"200"
               }
            ]
         }
      ],
      "values":[
         true
      ],
      "endpoint":"wss://rpc.rococo-parachain.litentry.io/"
   },
   "issuer":{
      "id":"did:litentry:substrate:0xdc9344ab8eca1af52e333670934a3b385cfd0e746ac876de63ad8726c83a878b",
      "name":"Litentry TEE Worker",
      "mrenclave":"7wipKrD5iVkPaHiAJC54aKRj5MVdwRUERHfuJRm5sLRY",
      "runtimeVersion":{
         "parachain":9180,
         "sidechain":106
      }
   },
   "issuanceDate":"2024-06-03T18:28:41.101850062+00:00",
   "parachainBlockNumber":27,
   "sidechainBlockNumber":39,
   "proof":{
      "created":"2024-06-03T18:28:41.102904979+00:00",
      "type":"Ed25519Signature2020",
      "proofPurpose":"assertionMethod",
      "proofValue":"201f80d0e574e442aeed764b0e5e803867b3837be1c40c4268a4bcd2a981cb91ab7b1a499ff1603e568d6ec3e8e0d9a58e69496ef678fefcabbeb6927db0b601",
      "verificationMethod":"0xdc9344ab8eca1af52e333670934a3b385cfd0e746ac876de63ad8726c83a878b"
   },
   "credentialSchema":{
      "id":"https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
      "type":"JsonSchemaValidator2018"
   }
}
```
Case 2 (false VC):
```
$ ./bin/litentry-cli -U wss://localhost/ trusted -d link-identity did:litentry:substrate:0xd4e35b16ec6b417386b948e7eaf5cc642a243096cecf366e6313689b90969f42 did:litentry:evm:0x83067bAE0D4468455948B87DEe5B2585de331B7D bsc
$ ./bin/litentry-cli -U wss://localhost/ trusted -d request-vc did:litentry:substrate:0xd4e35b16ec6b417386b948e7eaf5cc642a243096cecf366e6313689b90969f42 -a "dynamic 0000000000000000000000000000000000000001"
>>> identity: Substrate(Address32)
>>> nonce: 2
>>> assertions: [Dynamic(0x0000000000000000000000000000000000000001)]
```
----Generated VC-----

```
{
   "@context":[
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1"
   ],
   "id":"0x833cd685184be6f9b060a95896d46172bc3289d47e9e4f1964fa8f19fc831eda",
   "type":[
      "VerifiableCredential"
   ],
   "credentialSubject":{
      "id":"did:litentry:substrate:0xd4e35b16ec6b417386b948e7eaf5cc642a243096cecf366e6313689b90969f42",
      "description":"The amount of a particular token you are holding",
      "type":"Token Holding Amount",
      "assertionText":"Dynamic(0x0000000000000000000000000000000000000001)",
      "assertions":[
         {
            "and":[
               {
                  "src":"$token",
                  "op":"==",
                  "dst":"ordi"
               },
               {
                  "src":"$holding_amount",
                  "op":">=",
                  "dst":"0"
               },
               {
                  "src":"$holding_amount",
                  "op":"<",
                  "dst":"1"
               }
            ]
         }
      ],
      "values":[
         false
      ],
      "endpoint":"wss://rpc.rococo-parachain.litentry.io/"
   },
   "issuer":{
      "id":"did:litentry:substrate:0xdc9344ab8eca1af52e333670934a3b385cfd0e746ac876de63ad8726c83a878b",
      "name":"Litentry TEE Worker",
      "mrenclave":"7wipKrD5iVkPaHiAJC54aKRj5MVdwRUERHfuJRm5sLRY",
      "runtimeVersion":{
         "parachain":9180,
         "sidechain":106
      }
   },
   "issuanceDate":"2024-06-03T18:30:31.394300151+00:00",
   "parachainBlockNumber":36,
   "sidechainBlockNumber":58,
   "proof":{
      "created":"2024-06-03T18:30:31.395068879+00:00",
      "type":"Ed25519Signature2020",
      "proofPurpose":"assertionMethod",
      "proofValue":"7d0f07072e4b4706b9d92986493117eb39323db81058e91763b714d0170deb6b2e251aea5719208b385ceba850d8c1b958be868940d84cdc6460b55a2f694d0e",
      "verificationMethod":"0xdc9344ab8eca1af52e333670934a3b385cfd0e746ac876de63ad8726c83a878b"
   },
   "credentialSchema":{
      "id":"https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
      "type":"JsonSchemaValidator2018"
   }
}
```

kziemianek commented on tee-worker/litentry/core/evm-dynamic-assertions/src/precompiles/string_to_number.rs:
```diff
@@ -0,0 +1,174 @@
+use rust_decimal::prelude::{Decimal, FromStr};
+use std::{format, vec::Vec};
+
+pub fn string_to_number(input: Vec<u8>) -> PrecompileResult {
```
What do you think about replacing this with three precompiles:
* `parse_int`
* `parse_decimal`
* `hex_to_number`

The interface would be simpler and it would be easier to extend it in the future without introducing breaking changes.

wli-pro replied:
This is resolved now

kziemianek reviewed (commented):
I'll continue the review later. I'm sharing my current thoughts

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/A20.sol:
```diff
@@ -44,7 +37,9 @@ contract A20 is DynamicAssertion {
 
                 string memory url = concatenateStrings(
-                    "http://localhost:19527/events/does-user-joined-evm-campaign?account=",
+                    "https://api.twitter.com/events/does-user-joined-evm-campaign?account=",
```
This is not twitter api 😂

wli-pro replied:
Fixed now

kziemianek commented on tee-worker/litentry/core/evm-dynamic-assertions/src/precompiles/identity_to_string.rs:
```diff
@@ -0,0 +1,272 @@
+	Ok(bytes.to_base58())
+}
+
+fn web3_network_to_chain(network: u32) -> String {
```
We already have such mapping in `network_to_token`, maybe it can be reused instead ?

wli-pro replied:
network_to_token is sort of reverse of web3_network_to_chain though:
network_to_token: convert Web3Network type to numeric network code
web3_network_to_chain: convert numeric network code to Web3Network name string

0xverin commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/A20.sol:
```diff
@@ -18,22 +18,15 @@
-    {
+    function execute(
+        Identity[] memory identities,
+        string[] memory secrets
```
Parameters that need to be passed but are not used can be written like this: `string[] memory /*secrets*/`
Try to eliminate compilation warnings, otherwise there may be more and more later.

wli-pro replied:
Fixed now

kziemianek commented:
Please add precompile tests similar to ones added in #2786

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/DynamicAssertion.sol:
```diff
@@ -113,40 +83,72 @@ abstract contract DynamicAssertion {
+        uint256 encoded_len = encoded.length;
+
+        assembly {
+            if iszero(call(not(0), 0x03EA, 0, add(encoded, 0x20), encoded_len, returnVal, 0x82)) {
```
`0x32` means that only 32 bytes of returned string will be read from result of this call, remaining will be omitted

kziemianek commented on tee-worker/litentry/core/assertion-build/src/dynamic/contracts/DynamicAssertion.sol:
```diff
@@ -113,40 +83,72 @@ abstract contract DynamicAssertion {
+
+        assembly {
+            if iszero(call(not(0), 0x03EA, 0, add(encoded, 0x20), encoded_len, returnVal, 0x82)) {
+                revert(0, 0)
```
Please check how i'm handling string result from `toHex` precompile in this commit:
https://github.com/litentry/litentry-parachain/blob/de51144efb50c910c09dfa63b5a2403a9156ae09/tee-worker/litentry/core/assertion-build/src/dynamic/contracts/DynamicAssertion.sol#L131-L158

We should always advance free memory pointer located under `0x40` address otherwise bytes will be overriden by following memory allocations.

I've also removed stripping first 32bytes of encoded string in toHex precompile see: tee-worker/litentry/core/evm-dynamic-assertions/src/precompiles/to_hex.rs

If that's ok please follow so we have consistent impl.

BillyWooo commented on tee-worker/litentry/core/assertion-build/src/dynamic/mod.rs:
```diff
@@ -225,6 +226,53 @@ pub mod tests {
 		assert!(!credential.credential_subject.values[0]);
 	}
 
+	#[test]
```
Here only add ORDI as example to represent all other tokens?

Commit:
feat: P-701 solidity VC: token holding amount for BRC20 tokens
1) add the precompile function: test_hex_to_number, integration_test, parse_decimal, parse_int
2) extract new common functions for Web3Network: get_code, from_code, get_name
3) move integration test of precompile function to its own file
4) move all util functions to new library files
5) add BRC20 contracts, one contract only support one token
6) add new library AssertionLogic instead of assemble assertion JSON manually

kziemianek commented on tee-worker/litentry/core/evm-dynamic-assertions/src/precompiles/parse_int.rs:
```diff
@@ -0,0 +1,133 @@
+	use crate::{execute_smart_contract, prepare_function_call_input};
+	use ethabi::{decode, encode, ethereum_types::U256, ParamType, Token};
+
+	// tee-worker/litentry/core/assertion-build/src/dynamic/contracts/tests/ParseInt.sol
```
I was thiniking about where to place these tests and I decided to put it in package where smart contracts are located simply becuase the goal of these tests is to prove that solidity libraries works as expected (precompiled called and result properly returned) but that's fine too.

kziemianek reviewed (approved):
Great refactorings and improvements.
